File: LeetCode - Top Interview 150/2.py
# 2. Add Two Numbers
# https://leetcode.com/problems/add-two-numbers/


from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        dummy = current = ListNode()

        l1_ptr = l1
        l2_ptr = l2

        has_carry = None

        while l1_ptr or l2_ptr:
            value = None

            if l1_ptr and not l2_ptr:
                value = l1_ptr.val
                l1_ptr = l1_ptr.next

            elif not l1_ptr and l2_ptr:
                value = l2_ptr.val
                l2_ptr = l2_ptr.next

            else:
                value = l1_ptr.val + l2_ptr.val
                l1_ptr = l1_ptr.next
                l2_ptr = l2_ptr.next

            if has_carry:
                value += 1
                has_carry = False

            if value >= 10:
                value -= 10
                has_carry = True

            current.next = ListNode(value, None)
            current = current.next

        if has_carry:
            current.next = ListNode(1, None)

        return dummy.next

File: LeetCode JS,TS/2.ts
// 2. Add Two Numbers
// https://leetcode.com/problems/add-two-numbers/

// time complexity: O(max(m, n)) || space complexity: O(max(m, n))
function addTwoNumbers(
  l1: ListNode | null,
  l2: ListNode | null
): ListNode | null {
  if (!l1 || !l2) return null;

  let l1Pointer: ListNode | null = l1;
  let l2Pointer: ListNode | null = l2;

  let result: ListNode = new ListNode();
  let resultPointer: ListNode = result;

  let carry = 0;
  while (true) {
    let l1Value = l1Pointer?.val ? l1Pointer?.val : 0;
    let l2Value = l2Pointer?.val ? l2Pointer?.val : 0;

    const remainder = (l1Value + l2Value + carry) % 10;
    const quotient = (l1Value + l2Value + carry - remainder) / 10;

    resultPointer.val = remainder;
    carry = quotient;

    if (l1Pointer?.next || l2Pointer?.next || carry > 0) {
      resultPointer.next = new ListNode();
      resultPointer = resultPointer.next;

      l1Pointer = l1Pointer?.next ? l1Pointer.next : null;
      l2Pointer = l2Pointer?.next ? l2Pointer.next : null;
    } else {
      break;
    }
  }

  return result;
}

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}
